File: current_session/22.py
"""
see https://leetcode.com/problems/generate-parentheses/
"""
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        dp = [[] for _ in range(n+1)]   # dp[i]: output of i as n
        dp[0] += '',
        for i in range(n+1):
            for j in range(i):
                # add some inside, and add the others outside
                dp[i] += ['('+x+')'+y for x in dp[j] for y in dp[i-j-1]]
        return dp[n]

print(Solution().generateParenthesis(0))
print(Solution().generateParenthesis(1))
print(Solution().generateParenthesis(2))
print(Solution().generateParenthesis(3))
print(Solution().generateParenthesis(4))

File: current_session/c++/22.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

// DP
// dp[n] = (x)+y for all x+y=n-1
class Solution {
public:
    vector<string> generateParenthesis(int n) {
      if (n == 1) return vector<string>{"()"};
      // use dp
      vector<vector<string> > dp;
      dp.push_back(vector<string>{""});
      dp.push_back(vector<string>{"()"});
      for (int i=2; i<=n; i++) {
        // carefully generate x and y
        vector<string> tmp;
        for (int a=0; a<i; ++a) {
          int b = i-a-1;
          for (auto x: dp[a]) {
            for (auto y: dp[b]) {
              tmp.push_back("("+x+")"+y);
            }
          }
        }
        dp.push_back(tmp);
      }

      return dp[n];
    }
};